File: src/Bench.js
// performance.now is monotonic and expressed in milliseconds. Do not fall back
// to the adjustable wall clock when the host lacks this API.
const clock = globalThis.performance;
if (!clock || typeof clock.now !== "function") {
  throw new Error("Benchmark requires a monotonic performance.now clock");
}
const origin = clock.now();
export const benchNow = () => (clock.now() - origin) * 1000;
// This is the language-level Effect boundary; JS has no portable no-inline API.
const inputCell = new Int32Array(new SharedArrayBuffer(4));
const resultCell = new Int32Array(new SharedArrayBuffer(4));
export const opaque = (a) => () => {
  if (typeof a !== "number" || !Number.isInteger(a)) return a;
  Atomics.store(inputCell, 0, a);
  return Atomics.load(inputCell, 0);
};
export const formatNumber = (n) => n.toFixed(6);
export const measureBatch = (iterations) => (expected) => (act) => () => {
  let result = 0;
  const start = benchNow();
  for (let i = 0; i < iterations; i++) {
    result = act();
    Atomics.store(resultCell, 0, result);
  }
  const elapsed = benchNow() - start;
  if (result !== expected) throw new Error(`Unstable benchmark result: ${result} != ${expected}`);
  return elapsed / iterations;
};

File: src/Bench.java
public static final java.util.function.Supplier<Object> benchNow = () -> {
    return (double) (System.nanoTime() / 1000.0);
};

private static volatile Object benchmarkInput;
private static volatile Object benchmarkResult;
public static final java.util.function.Function<Object, Object> opaque = (a) ->
    (java.util.function.Supplier<Object>) () -> {
        benchmarkInput = a;
        return benchmarkInput;
    };

public static final java.util.function.Function<Object, Object> formatNumber = (nObj) -> {
    double n = ((Number) nObj).doubleValue();
    return String.format(java.util.Locale.US, "%.6f", n);
};

// The common numeric harness protects compiled and FFI kernels equally.
public static java.util.function.Function<Object, Object> nativeBenchmark(
        java.util.function.Function<Object, Object> action) {
    return action;
}

public static final java.util.function.Function<Object, Object> measureBatch = iterations ->
    (java.util.function.Function<Object, Object>) expected ->
    (java.util.function.Function<Object, Object>) action ->
    (java.util.function.Supplier<Object>) () -> {
        java.util.function.Supplier<?> act = (java.util.function.Supplier<?>) action;
        int count = ((Number) iterations).intValue();
        Object result = null;
        long start = System.nanoTime();
        for (int i = 0; i < count; i++) {
            result = act.get();
            benchmarkResult = result;
        }
        double elapsed = (System.nanoTime() - start) / 1000.0;
        if (((Number) result).intValue() != ((Number) expected).intValue())
            throw new IllegalStateException("Unstable benchmark result");
        return elapsed / count;
    };
